src/backend/dx12/src/native.rs

```diff
@@ -497,8 +574,68 @@ impl HalDescriptorPool<Backend> for DescriptorPool {
         })
     }
 
+    fn free_sets(&mut self, descriptor_sets: &[DescriptorSet]) {
+        for descriptor_set in descriptor_sets {
+            for binding_info in &descriptor_set.binding_infos {
+                if let Some(ref view_range) = binding_info.view_range {
+                    let HeapProperties { has_view, .. } = view_range.ty.heap_properties();
+                    if has_view {
+                        self.heap_srv_cbv_uav.free_handles(view_range.handle);
+                    }
+                    
+                }
+                if let Some(ref sampler_range) = binding_info.sampler_range {
+                    let HeapProperties { has_sampler, .. } = sampler_range.ty.heap_properties();
+                    if has_sampler {
+                        self.heap_sampler.free_handles(sampler_range.handle);
+                    }
+                }
+            }
+        }
+    }
+
     fn reset(&mut self) {
-        unimplemented!()
+        self.heap_srv_cbv_uav.clear();
+        self.heap_sampler.clear();
+    }
+}
+
+struct HeapProperties {
+    has_view: bool,
+    has_sampler: bool,
+    is_uav: bool,
+}
+
+trait HasHeapProperties {
```

Comment:
sorry, I didn't expect us to come up with an internal trait here (wasn't sure if the types are from this crate - they are not).
mind if we go back to just having a method then? Perhaps, `HeapProperties::from(DescriptorType)`?